homeassistant/components/vacuum/xiaomi.py

```diff
@@ -0,0 +1,373 @@
+            DOMAIN, vacuum_service, async_service_handler,
+            description=descriptions.get(vacuum_service), schema=schema)
+
+
+class MiroboVacuum(VacuumDevice):
+    """Representation of a Xiaomi Vacuum cleaner robot."""
+
+    def __init__(self, hass, name, host, token):
+        """Initialize the Xiaomi vacuum cleaner robot handler."""
+        self.hass = hass
+        self._name = name
+        self._icon = ICON
+        self._host = host
+        self._token = token
+        self._vacuum = None
+
+        self.vacuum_state = None
+        self._is_on = False
+        self._available = False
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon to use for device."""
+        return self._icon
+
+    @property
+    def state(self) -> str:
+        """Return the state of the vacuum cleaner as a binary state."""
+        return STATE_ON if self.is_on else STATE_OFF
+
+    @property
+    def status(self):
+        """Return the status of the vacuum cleaner."""
+        if self.vacuum_state is not None:
+            return self.vacuum_state.state
+
+    @property
+    def battery_level(self):
+        """Return the battery level of the vacuum cleaner."""
+        if self.vacuum_state is not None:
+            return self.vacuum_state.battery
+
+    @property
+    def fanspeed(self):
+        """Return the fan speed of the vacuum cleaner."""
+        if self.vacuum_state is not None:
+            speed = self.vacuum_state.fanspeed
+            if speed in FANSPEEDS.values():
+                return [key for key, value in FANSPEEDS.items()
+                        if value == speed][0]
+            return speed
+
+    @property
+    def fanspeed_list(self):
+        """Get the list of available fan speed steps of the vacuum cleaner."""
+        return list(FANSPEEDS.keys())
+
+    @property
+    def device_state_attributes(self):
+        """Return the specific state attributes of this vacuum cleaner."""
+        if self.vacuum_state is not None:
+            attrs = {
+                'Do not disturb':
+                    STATE_ON if self.vacuum_state.dnd else STATE_OFF,
+                # Not working --> 'Cleaning mode':
+                #    STATE_ON if self.vacuum_state.in_cleaning else STATE_OFF,
+                'Cleaning time': str(self.vacuum_state.clean_time),
+                'Cleaned area': self.vacuum_state.clean_area}
+            if self.vacuum_state.got_error:
+                attrs['Error'] = self.vacuum_state.error
+            return attrs
+
+        return {}
+
+    @property
+    def vacuum(self):
+        """Property accessor for vacuum object."""
+        if not self._vacuum:
+            from mirobo import Vacuum
+            _LOGGER.info("Initializing with host %s (token %s...)",
+                         self._host, self._token[:5])
+            self._vacuum = Vacuum(self._host, self._token)
```

Comment:
If this does I/O it should be moved to a method eg `async_update`. Doing this type of thing in a property is not common anyhow in home assistant. Since `async_update` will be called immediately after adding the entity, I suggest moving it there either way, if there's a reason not to put this in `__init__` other than I/O.